src/funcx_common/task_storage/s3.py

```diff
@@ -0,0 +1,50 @@
+import typing as t
+
+import boto3
+
+from ..tasks import TaskProtocol
+from .base import StorageException, TaskStorage
+
+
+class S3TaskStorage(TaskStorage):
+    """
+    Store Task data in S3
+
+    """
+
+    storage_id = "MemoryTaskStorage"
+
+    def __init__(self, bucket_name: str) -> None:
+        self.bucket_name = bucket_name
+        self.client = boto3.client("s3")
+        self._results: t.Dict[str, str] = {}
+
+    def store_result(
+        self, task: TaskProtocol, result: str, ACL: str = "authenticated"
+    ) -> bool:
+        key = f"{task.task_id}.result"
+        self.client.put_object(  # ACL=ACL,
+            Body=result.encode("utf-8"),
+            Bucket=self.bucket_name,
+            Key=key,
+        )
+        task.result_reference = {
+            "storage_id": self.storage_id,
+            "s3bucket": self.bucket_name,
+            "key": key,
+        }
+        return True
+
+    def get_result(self, task: TaskProtocol) -> t.Optional[str]:
+        if (
+            task.result_reference
+            and task.result_reference["storage_id"] == self.storage_id
+        ):
+            response = self.client.get_object(
+                Bucket=task.result_reference["s3bucket"],
+                Key=task.result_reference["key"],
+            )
+            # mypy seems to think that the following line does not return a string
+            return response["Body"].read().decode("utf-8")  # type: ignore
```

Comment:
When mypy isn't able to correctly deduce a type, it's generally preferable to use `typing.cast` to a type-ignore.
Ignoring can hide all kinds of things.

For example, if I write
```python
import typing as t

...

x = []
return t.cast(str, x["foo"])
```
then mypy can still flag this as a type error because it can still examine the type of `x`. With a type ignore, the whole line is missed by the checker.

Which stage of this is `mypy` unsure about? I'd tend to guess that it sees `response` as `Dict[str, Any]` and therefore is upset by the fact that this is returning `Any` from a method annotated as returning an optional string.
I think this is a good case for a cast to coerce the type to the correct thing.
In the future, boto3 may provide a TypedDict which allows correct inference in this case, and we could remove the cast.

Reply:
I've done a cast.